sw/host/opentitanlib/src/transport/hyperdebug/dfu.rs

```diff
@@ -0,0 +1,407 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Firmware update protocol for HyperDebug
+
+use anyhow::{bail, Context, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde_annotate::Annotate;
+use std::any::Any;
+use std::cell::RefCell;
+
+use crate::transport::{Capabilities, Capability, Transport, TransportError, UpdateFirmware};
+use crate::util::usb::UsbBackend;
+
+const VID_ST_MICROELECTRONICS: u16 = 0x0483;
+const PID_DFU_BOOTLOADER: u16 = 0xdf11;
+
+const USB_DFU_DETACH: u8 = 0;
+const USB_DFU_DNLOAD: u8 = 1;
+const USB_DFU_GETSTATUS: u8 = 3;
+
+/// This transport is to be used if a Nucleo board is already in DFU bootloader mode at the time
+/// of the `opentitantool` invocation (and presenting itself with STMs VID:DID, rather than
+/// Google's).
+pub struct HyperdebugDfu {
+    usb_backend: RefCell<UsbBackend>,
+}
+
+impl HyperdebugDfu {
+    /// Establish connection with a particular Nucleo-L552ZE board already in DFU bootloader mode.
+    pub fn open(
+        usb_vid: Option<u16>,
+        usb_pid: Option<u16>,
+        usb_serial: Option<&str>,
+    ) -> Result<Self> {
+        let usb_backend = UsbBackend::new(
+            usb_vid.unwrap_or(VID_ST_MICROELECTRONICS),
+            usb_pid.unwrap_or(PID_DFU_BOOTLOADER),
+            usb_serial,
+        )?;
+        Ok(Self {
+            usb_backend: RefCell::new(usb_backend),
+        })
+    }
+}
+
+/// The device does not support any part of the Transport trait, except the UpdateFirmware action.
+impl Transport for HyperdebugDfu {
+    fn capabilities(&self) -> Result<Capabilities> {
+        Ok(Capabilities::new(Capability::NONE))
+    }
+
+    fn dispatch(&self, action: &dyn Any) -> Result<Option<Box<dyn Annotate>>> {
+        if let Some(update_firmware_action) = action.downcast_ref::<UpdateFirmware>() {
+            update_firmware(
+                &mut self.usb_backend.borrow_mut(),
+                &update_firmware_action.firmware,
+                &update_firmware_action.progress,
+            )
+        } else {
+            bail!(TransportError::UnsupportedOperation)
+        }
+    }
+}
+
+const USB_CLASS_APP: u8 = 0xFE;
+const USB_SUBCLASS_DFU: u8 = 0x01;
+
+const DFUSE_ERASE_PAGE: u8 = 0x41;
+const DFUSE_PROGRAM_PAGE: u8 = 0x21;
+
+const DFU_STATUS_OK: u8 = 0x00;
+
+const DFU_STATE_APP_IDLE: u8 = 0x00;
+const DFU_STATE_DFU_IDLE: u8 = 0x02;
+const DFU_STATE_DOWNLOAD_BUSY: u8 = 0x04;
+const DFU_STATE_DOWNLOAD_IDLE: u8 = 0x05;
+
+fn official_firmware_image() -> &'static [u8] {
+    include_bytes!(env!("hyperdebug_firmware"))
+}
```

Comment:
I also would prefer not to bundle the firmware in the opentitantool binary.

Reply:
There were two reasons I did it this way:
1) It conveniently gives the `HyperDebug::open()` method access to the textual version string of the "official" firmware, which it compares against that of the running firmware, and print a warning message, in case the debugger is not running the latest.
2) The Chrome OS automatic test lab uses Docker images containing the test tools to be used.  We construct one such Docker image by checking out the OpenTitan git repository, compiling opentitantool, and putting the resulting binary along with configuration files, and other Chrome OS scripts and binaries into the testing image.  Notably, bazel is NOT part of the testing image.  That means that on our test systems, we are limited to running the binary `opentitantool`, and it would still be convenient to be able to invoke `opentitantoo transport update-firmware` and be certain that the debugger hardware then would have a firmware exactly compatible with the rust code in the `opentitantool` binary in the testing image.

Sure, we could work around the second issue by also including the firmware image in our Docker image. Under your suggestion, Bazel should be able to download it, and put it in the `out` or `bin` directory along with the opentitantool executable, but it would still be more steps for us.

Also, in order to retain the ability for opentitantool to alert the user to firmware version mismatch, at least the version string needs to be embedded into the opentitantool binary.  I am sure it can be done, but it is some work to create another little rust program, which extracts the version from a HyperDebug image file, and make that part of the build process, such that Bazel will download the firmware image, extract the version, and compile the version, rather than the entire 64kB image, into opentitantool.

Reply:
As per our discussion, I have made inclusion of the HyperDebug firmware optional, controlled by a Rust feature flag, `include_hyperdebug_firmware`.